feat(web): View all duplicates page

arnolicious opened the pull request:
## Feature Description
Add a new intermediate page, that displays all the found duplicates alongside how many duplicate assets each duplicate has.
From there, you can click on any duplicate to arrive at the "old" duplicate selection mechanism for that specific stack.
After resolving, it should move on to the next duplicate in the list
![image](https://github.com/user-attachments/assets/70443f98-fda0-4ff7-aaaa-83ea4d8c0f74)

With this change, we can now also link from "Viewing an asset" straight to the duplication page for that duplicate stack.

![image](https://github.com/user-attachments/assets/b8dcef6a-82c3-4a08-bcad-e36de0e448e1)

## Implementation
- Changed the page params of the "old" duplicate page to require the `duplicateId`
  - Here I'm a little unsure why there's the optional `[[photos=photos]]` param, I assume it's not relevant anymore with these changes, but I wanted confirmation before removing it
- Added the new page under `utilities/duplicates`
- Added a link to the deduplication page from the asset viewer, if the asset has a `duplicateId`

arnolicious committed:
feat: added a page to view all duplicates

arnolicious committed:
jump to the next duplicate in the queue

arnolicious committed:
added link from asset viewer to duplicate page

arnolicious committed:
Merge branch 'main' into feat/duplicates-grid-view

bo0tzz commented:
Can you add some screenshots to the PR description?

arnolicious commented:
There you go :) @bo0tzz

arnolicious committed:
removed unused import

alextran1502 committed:
Merge branch 'main' into feat/duplicates-grid-view

alextran1502 commented:
What happens if you have 10_000 or 20_000 duplicates? Will it crash the page?

arnolicious commented:
Seems to run just fine, at least for me. I adjusted the load function of `utilities/duplicates/+page.ts` like this:

```ts
export const load = (async ({ params }) => {
  await authenticate();
  const asset = await getAssetInfoFromParam(params);
  const duplicates = await getAssetDuplicates();
  const $t = await getFormatter();

  const mockedDuplicates = [...duplicates];

  for (let i = 0; i < 10_000; i++) {
    mockedDuplicates.push(...duplicates);
  }

  return {
    asset,
    duplicates: mockedDuplicates,
    meta: {
      title: $t('duplicates'),
    },
  };
}) satisfies PageLoad;

```

arnolicious committed:
handle case where no duplicates are left

arnolicious committed:
Merge branch 'feat/duplicates-grid-view' of https://github.com/arnolicious/immich into feat/duplicates-grid-view

alextran1502 commented:
How about 20,000? usually, with these types of rendering, the load can be stressed on both the server load and the web browser crashing. I am very positive we will run into issues since the query isn't lazy paginated as well as the view doesn't have lazy loading mechanism

I chose to go with single duplicate because of the aforementioned limitation

arnolicious commented:
Alright, so we'd need to implement a client side pagination to merge this?

I assume only client side, since the current single duplicate implementation also already calls the entire duplicate list from the server, so this shouldn't be a problem

alextran1502 commented:
Let's take a few steps and understand the problem we are trying to solve. Below are some of my feed back, and thoughts

- The intermediate page doesn't have a clear indication that the user can click on the group of duplicates to handle them.
- There should be bulk action on this page, and the group of duplicate should be selectable to perform deduplication.
- I agree with you that we should have pagination on this page, breadcrumb navigation is a good solution.
- It would be nice to show all the duplicated assets while hovered on each group

Heck, perhaps we can replace the current view if the above points are implemented. What do you think?

NicholasFlamy commented:
Make sure to consider #13851 before you get too far.

mertalev commented on web/src/lib/components/utilities-page/duplicates/duplicate-thumbnail.svelte:
```diff
@@ -0,0 +1,42 @@
+  let { duplicate }: Props = $props();
+
+  let assetToDisplay = duplicate.assets[0];
+  let title = $derived(
+    JSON.stringify(
+      duplicate.assets.map((asset) => asset.originalFileName),
+      null,
+      2,
+    ),
+  );
```
It should just be something like `${assetToDisplay.originalFilename} duplicates`

mertalev commented on web/src/lib/components/asset-viewer/detail-panel.svelte:
```diff
@@ -464,6 +465,23 @@
+              href="{AppRoute.DUPLICATES}/{asset.duplicateId}"
+              class="hover:dark:text-immich-dark-primary hover:text-immich-primary"
+            >
+              {$t('has_duplicates')}
```
I feel this doesn't give a good indication that it's a button that takes you to a different view.

NicholasFlamy replied:
I like this feature to see the duplicate info on the image details. I agree with mert about it not being clear that it is clickable.

arnolicious replied:
Tried to improve this in the latest iteration :D

mertalev commented on web/src/routes/(user)/utilities/duplicates/[[photos=photos]]/[duplicateId=id]/+page.ts:
```diff
@@ -0,0 +1,25 @@
+
+export const load = (async ({ params }) => {
+  await authenticate();
+  const duplicates = await getAssetDuplicates();
```
This view only needs to know the number of each duplicate group, not fetch all their members. It can fetch much less data with a different endpoint that only returns the data it actually needs (one asset to display per group, and the number of duplicates in the group).

mertalev commented on web/src/routes/(user)/utilities/duplicates/[[photos=photos]]/[duplicateId=id]/+page.svelte:
```diff
@@ -103,7 +116,17 @@
     await stackAssets(assets, false);
     const duplicateAssetIds = assets.map((asset) => asset.id);
     await updateAssets({ assetBulkUpdateDto: { ids: duplicateAssetIds, duplicateId: null } });
+    const currentDuplicateIndex = duplicates.findIndex((duplicate) => duplicate.duplicateId === duplicateId);
```
Can't you just track the index instead of doing this?

arnolicious replied:
Sorry, I don't think I really understand by "tracking" in this context.
Where would you calculate the index? In the API endpoint?

midzelis replied:
Maybe what they meant was to pass in the index to `handleResolve` instead of finding it?

mertalev commented on web/src/routes/(user)/utilities/duplicates/[[photos=photos]]/[duplicateId=id]/+page.ts:
```diff
@@ -0,0 +1,25 @@
+  const duplicates = await getAssetDuplicates();
+  const $t = await getFormatter();
+
+  const activeDuplicate = duplicates.find((duplicate) => duplicate.duplicateId === params.duplicateId);
```
Here as well.

arnolicious commented:
Alright, thanks for all the feedback! I'll give it a try in the coming days/weeks

My idea for combining the 2 views would be something like this:
Having the duplicates group be selectable, and when selected the old duplicate resolution component gets shown at the top of the page.

Is that something you had in mind?

alextran1502 commented:
I think it is worth to draft a wireframe design so we don't misunderstand how the page will look like

arnolicious commented:
How about something like this?
The currently selected group would be the one displayed in the top section, and visible in the bottom list with a border.
![image](https://github.com/user-attachments/assets/1ba44716-6984-4fa5-bb71-1249354f9612)

Tho for my personal use case, I would wish to be able to see a bigger part of all the duplicates somewhere, somehow

mertalev commented:
I'm not a huge fan of the explicit numbered pages. Maybe there's a more subtle way to paginate this, or to use virtual scroll like the timeline?

arnolicious commented:
I did also consider this, would just be a bit more clicking if you specifically wanted to get to the end of the list for example
![image](https://github.com/user-attachments/assets/b981dfa0-54e8-49bb-b8d4-abf278250cc2)

alextran1502 commented:
> I did also consider this, would just be a bit more clicking if you specifically wanted to get to the end of the list for example ![image](https://private-user-images.githubusercontent.com/46051866/410994311-b981dfa0-54e8-49bb-b8d4-abf278250cc2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg5NzEzMDksIm5iZiI6MTczODk3MTAwOSwicGF0aCI6Ii80NjA1MTg2Ni80MTA5OTQzMTEtYjk4MWRmYTAtNTRlOC00OWJiLWI4ZDQtYWJmMjc4MjUwY2MyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjA3VDIzMzAwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNkMjU3ODQxNDVkMzhmMWZkOGE4YTlhY2ZjNjk1NWM2NGU2ZDUyZGM3Nzk1ZTNmZTRhNjEyYjI0NjY0ZWEzZDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.1GgBbysrM8b7XDQPslupbNqiDiirRENQk_S_Vs2c4oM)

I like this a lot better, how are you planning to handle loading, paginating in this scroll view?

NicholasFlamy commented:
> I did also consider this, would just be a bit more clicking if you specifically wanted to get to the end of the list for example ![image](https://private-user-images.githubusercontent.com/46051866/410994311-b981dfa0-54e8-49bb-b8d4-abf278250cc2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg5NzUwNDMsIm5iZiI6MTczODk3NDc0MywicGF0aCI6Ii80NjA1MTg2Ni80MTA5OTQzMTEtYjk4MWRmYTAtNTRlOC00OWJiLWI4ZDQtYWJmMjc4MjUwY2MyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjA4VDAwMzIyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY4MjViMDdhYzliZTNkZTIwZjNmNGQwZWJjZDk3NjA3Y2Y3YTc3Zjc0NDA0ZWU1NzFhMWViN2UyYjhkZWI3MzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.tkQqeC4uJV6ORsgmDSImMfjY3fc8VzDGHlDHlkymbtM)

I love this. This is exactly what I would like.

arnolicious commented:
I would create a `renderedItems` list that only contains the current page + the ones before and after.
And each time you change page, that list is updated like a sliding window over the entire list.

I don't think that should be too hard to implement, without the need for any additional libraries.

arnolicious committed:
Merge branch 'main' into feat/duplicates-grid-view

mertalev commented:
I agree a vertical scroll oriented design would be more intuitive, easier to use and more consistent with other pages.

That being said, I'm not a web dev and not sure how difficult it is to make it a virtual scroll. It's at least easy to do it the way the search result page works (load more results on scroll without removing earlier entries from the DOM). It would start having issues once you scroll enough to have >5000 groups loaded, but that might be fine tbh.

NicholasFlamy commented:
> That being said, I'm not a web dev and not sure how difficult it is to make it a virtual scroll.

Just a bit ago I ran into the work done for the timeline and I almost fainted. (The timeline has inconsistent sizing that has to be estimated and fixed on the scrollbar and stuff, so that's why it's so complicated.)

arnolicious commented:
Well tbh, this was more of a ui problem than a technical one. I'm sure we can make virtual scroll work horizontally or vertically.

I just didn't have a good idea on how to make a vertically scrolling list/grid while also having the dedupe mechanism on the same page.

If someone's got ideas or suggestions, I'm open :D

NicholasFlamy commented:
> on the same page

Is that truly necessary? What about clicking on an image pops up the current page (to the specific image you click)? And if you wanna be fancy, it could be a big (almost full-screen) popup.

arnolicious commented:
Well, that is the current implementation, basically. List of duplicate groups, you click on it and land on the known duplicate page.

But Alex suggested to merge those into one: https://github.com/immich-app/immich/pull/15856#issuecomment-2631906971

NicholasFlamy commented:
> Well, that is the current implementation, basically. List of duplicate groups, you click on it and land on the known duplicate page.

Oh yeah, I just tried it. It's pretty good but yeah, some stuff could be addressed. Also, you should try rebasing when you merge main.

arnolicious committed:
Merge remote-tracking branch 'origin/main' into feat/duplicates-grid-view

arnolicious committed:
added virtual list in duplicates view & more duplicates endpoints

arnolicious commented:
So, a little update on what I've done now:

- Added a virtual list to virtualize the duplicates grid which should avoid performance issues on huge lists
- Added the Bulk actions to the overview page
- Removed the bulk actions from the detail page, since that seemed confusing / redundant now
- Hopefully improved UX:
  - Added helper text to the grid view, to explain to click on a duplicate group to handle it 
  - Added a border when hovering over the items in the grid view to show "more clickability"

In wanting to implement Mert's comments, and improve effiency in general, I also added 2 new endpoints for duplicates:
- `GET duplicates/:id`
- `GET duplicates/info` - Gives out the list of all duplicates, but only with one asset per duplicate

However I didn't end up actually using those:
- Since the bulk actions are now in the grid view, I do need the entire asset list there to perform the bulk actions.
- The `duplicates/:id` Endpoint was meant for the detail page, however if I don't have the list of all the duplicates, idk how to load the next duplicate group after handling the current one.

arnolicious committed:
fixed the title of thumbnail

arnolicious committed:
make link to duplicate more understandable

zackpollard commented:
This feature looks great, for now I am going to convert the PR to a draft, once you're ready for a review again, please feel free to change it back once you've implemented the requested changes and then make new requests for reviews from people. Thanks!

arnolicious committed:
Merge branch 'main' into feat/duplicates-grid-view

arnolicious committed:
fixed lockfile

arnolicious committed:
fix: fixed new modal references

arnolicious committed:
fix: removed unnecessary type assertions

arnolicious committed:
fix: added type narrowing

arnolicious committed:
fix(web): fix web lint issues

arnolicious committed:
fix: regenerated openapi client

arnolicious committed:
fix: i18n formatting & ambiguous column reference

arnolicious committed:
fix: another ambiguous column reference

arnolicious committed:
fix: generated sql

arnolicious committed:
Merge branch 'main' into feat/duplicates-grid-view

danieldietzler reviewed (commented):
I am a bit worried that this won't really scale well in its current implementation. I think realistically we need to expect people to have >10k duplicates? Happy to hear opinions from the others

mertalev commented:
Fetching all duplicates is the status quo right now, so that in itself isn't a regression. Agree it should be tested on a large number of duplicates though.

arnolicious commented on web/package.json:
```diff
@@ -51,6 +51,7 @@
     "svelte-gestures": "^5.1.3",
     "svelte-i18n": "^4.0.1",
     "svelte-maplibre": "^1.0.0",
+    "svelte-virtual": "^1.0.0-next.22",
```
That library has introduced svelte 5 support in the `next.19` version, I have no clue why the author keeping it behind the `next flag`, even the readme's install instruction use the `next` version...

https://github.com/ghostebony/svelte-virtual

bo0tzz replied:
Should that dep be added at all, don't we already have our own existing implementations for this sort of thing in the web project?

danieldietzler replied:
@bo0tzz you really should learn to keep reading (in this case through other review comments) lmfao

bo0tzz replied:
I actually did this time but somehow managed to miss that one :P

arnolicious commented on web/src/routes/(user)/utilities/duplicates/+page.svelte:
```diff
@@ -0,0 +1,177 @@
+  let hasDuplicates = $derived(duplicates.length > 0);
+  let isShowingDuplicateInfo = $state(false);
+
+  let virtualGridContainer = $state<HTMLElement>();
+
+  const DEFAULT_HEIGHT = 780;
+
+  let virtualGridHeight = $state(DEFAULT_HEIGHT);
+
+  $effect(() => {
+    if (virtualGridContainer) {
+      virtualGridHeight = virtualGridContainer.clientHeight;
+    }
+  });
```
Yeah, tbf I did not even try to look at or use the `AssetGrid`, I kinda assumed it would be "overkill", since we aren't dealing with these different image sizes, month / day groupings etc...

I could try it out, if it is something that would also make sense here :D

danieldietzler replied:
I'm not too excited about having more usages of AssetGrid, but pulling in a new dependency for that is just kind of annoying. If it's so simple here wondering if we could just build it ourselves (or rip a small part of logic out of the AssetGri)

arnolicious replied:
Well actually, list and grid virtualization is something I'll have to deal with at work in the coming weeks, so I might aswell give it a go at building a solution "from scratch" here, instead of using `svelte-virtual` or `AssetGrid`...

midzelis replied:
[AssetGrid ](https://github.com/immich-app/immich/blob/main/web/src/lib/components/photos-page/asset-grid.svelte) would in fact probably be overkill. However, I don't think we need a new from-scratch implementation of yet another virtual-viewer. 

I'm planning some much overdue refactorings for asset-grid/asset-store. As part of this work, I was planning to extract a base out of asset-grid. 

[Asset-Grid](https://github.com/immich-app/immich/blob/main/web/src/lib/components/photos-page/asset-grid.svelte) is primarily a "set" viewer. It is virtual, and it "chunks" the total set that is being view. It supports scrubbing and instantaneous jumping to any point in the set. And yes, to perform this well, it subdivides the set into sub-sets, which happen to be aligned to months and days. These grouping do not need to be months/days - they could (in the future, after some appropriate refactoring) group things by location (city/state), by "event", by "tag", by "person" or any other grouping that makes sense. 

[Gallery-Viewer](https://github.com/immich-app/immich/blob/main/web/src/lib/components/shared-components/gallery-viewer/gallery-viewer.svelte) is primarily a "streaming results".  You can not jump (no scrubbing) - but you can scroll infinitely. As you scroll, you stream in more results from the data-source. 

[Memory-Viewer ](https://github.com/immich-app/immich/blob/main/web/src/lib/components/shared-components/gallery-viewer/gallery-viewer.svelte) (the part you see before "opening" the viewer itself) is very similar to gallery-viewer, except that is uses horizontal scrolling instead of vertical. It also doesn't share any of the gallery-viewer components. 

What you created here is more similar to gallery-viewer than to asset-grid. I like your enhancement. But, I think it would be a good opportunity to develop a common component that can support Gallery-Viewer/Memory-Viewer and this duplicates-viewer scenarios, to prevent to proliferation of various components that do almost-the-same things. 

This is a good start - but I would like to see `duplicate-grid.svelte` be turned into a generic component, and then used in gallery-viewer/memory-viewer as well. 

That means, it should support:
1) Horizontal or Vertical scrolling
2) Streaming data from the datasource as it gets close to the end of scroll
3) Gallery-Viewer actually uses justified-layout, so the component should be agnostic to the actual layout algorithms, i.e.  allow you to pass in a function for layout. 
4) Have UI hooks (snippets) for specialized controls (i.e. the memory viewer left/right buttons)

arnolicious committed:
feat: added diy grid virtualization with lazy loading images and placeholders when scrolling fast

arnolicious committed:
Merge branch 'main' into feat/duplicates-grid-view

arnolicious committed:
fix: replace alt text

arnolicious commented:
Alrighty, I have some improvements:
- The duplicate image elements now only fetch the image from the server, once it's in the viewport
- Built a DIY virtual grid that only renders items in the viewport, replacing the additional dependency
- It also tracks the scroll speed, to only render a placeholder when scrolling fast
  - This helps when scrolling far down faster: Since only the placeholders are rendered before, no img fetches are started, that would hang up the fetches of the currently visible images

arnolicious committed:
Merge branch 'main' into feat/duplicates-grid-view

arnolicious committed:
Merge branch 'main' into feat/duplicates-grid-view

midzelis commented:
> > I did also consider this, would just be a bit more clicking if you specifically wanted to get to the end of the list for example 
> 
> I like this a lot better, how are you planning to handle loading, paginating in this scroll view?

I also like this version a lot.

midzelis commented on web/src/lib/components/utilities-page/duplicates/duplicate-thumbnail.svelte:
```diff
@@ -0,0 +1,60 @@
+  {#if asPlaceholder}
+    <div class="h-60 bg-gray-200 dark:bg-gray-800 rounded-xl w-full animate-pulse"></div>
+  {:else}
+    <img alt={title} {title} class="h-60 object-cover rounded-xl w-full" draggable="false" bind:this={imgElement} />
```
you don't need title and alt at same time

midzelis commented on web/src/routes/(user)/utilities/duplicates/[[photos=photos]]/[duplicateId=id]/+page.svelte:
```diff
@@ -203,12 +153,12 @@
       </div>
 
-      {#key duplicates[0].duplicateId}
+      {#key activeDuplicate.duplicateId}
```
Can this #key be removed?

midzelis commented on web/src/lib/components/utilities-page/duplicates/duplicate-grid.svelte:
```diff
@@ -0,0 +1,170 @@
+
+    scrollTop = e.currentTarget.scrollTop;
+
+    isScrollingFast = scrollSpeed > IS_SCROLLING_FAST_THRESHOLD;
```
Also, if you were calculating scrolling speed, I think you'd have to measure the distance traveled / time -- a lot of browsers don't onscroll events at the same rate, especially if i.e. low on power, or high cpu usage, etc

arnolicious commented on web/src/routes/(user)/utilities/duplicates/+page.svelte:
```diff
@@ -0,0 +1,180 @@
+    bind:this={virtualGridContainer}
+    class="mt-2 h-[calc(100%-theme(spacing.20))] overflow-auto immich-scrollbar"
+  >
+    <DuplicateGrid items={duplicates} itemHeight={256} itemWidth={176} containerHeight={virtualGridHeight}>
```
Are there other places where there are such hard-coded image dimensions?

zackpollard commented:
Hey, I am going to close this for now due to it being stale for quite a long time. If you do get time to make the requested changes please feel free to make a new PR or comment to have this one re-opened :slightly_smiling_face: 

Also as a side note, personally I am not a huge fan of us implementing an entirely custom grid just for this page, this page doesn't feel that special to me.
